packages/gensx/src/dev-server.ts

```diff
@@ -0,0 +1,1192 @@
+
+        if ("name" in workflow && typeof workflow.name === "string") {
+          workflowName = workflow.name;
+        } else {
+          // Fallback to export name if name not found
+          workflowName = exportName;
+        }
+
+        this.workflowMap.set(workflowName, workflow);
+        //console.info(`Registered workflow: ${workflowName}`);
+      }
+    }
+
+    if (this.workflowMap.size === 0) {
+      console.warn("⚠️ No valid workflows were registered!");
+    }
+  }
+
+  /**
+   * Get a workflow by name or throw NotFoundError
+   */
+  private getWorkflowOrThrow(workflowName: string): unknown {
+    const workflow = this.workflowMap.get(workflowName);
+    if (!workflow) {
+      throw new NotFoundError(`Workflow '${workflowName}' not found`);
+    }
+    return workflow;
+  }
+
+  /**
+   * Get an execution by ID or throw NotFoundError
+   */
+  private getExecutionOrThrow(
+    executionId: string,
+    workflowName: string,
+  ): WorkflowExecution {
+    this.getWorkflowOrThrow(workflowName);
+    const execution = this.executionsMap.get(executionId);
+    if (!execution) {
+      throw new NotFoundError(`Execution '${executionId}' not found`);
+    }
+
+    if (workflowName && execution.workflowName !== workflowName) {
+      throw new NotFoundError(
+        `Execution '${executionId}' does not belong to workflow '${workflowName}'`,
+      );
+    }
+
+    return execution;
+  }
+
+  /**
+   * Parse request body with error handling
+   */
+  private async parseJsonBody(c: Context): Promise<Record<string, unknown>> {
+    try {
+      return await c.req.json();
+    } catch (_) {
+      throw new BadRequestError("Invalid JSON");
+    }
+  }
+
+  /**
+   * Validate input against schema
+   * Throws BadRequestError if validation fails
+   */
+  private validateInput(workflowName: string, input: unknown): void {
+    // Check if input is missing
+    if (input === undefined) {
+      throw new BadRequestError(
+        "Missing required 'input' field in request body",
+      );
+    }
+
+    // Get schema for this workflow
+    const schema = this.schemaMap.get(workflowName);
+    if (!schema?.input) {
+      // If no schema, we can't validate
+      return;
+    }
+
+    // Use Ajv to validate the input against the schema
+    const validate = this.ajv.compile(schema.input);
+    const valid = validate(input);
+
+    if (!valid) {
+      const errors = validate.errors ?? [];
+      const errorMessages = errors
+        .map((err: ErrorObject) => `${err.instancePath} ${err.message}`)
+        .join("; ");
+
+      throw new BadRequestError(
+        `Input validation failed: Input${errorMessages}`,
+      );
+    }
+  }
+
+  /**
+   * Set up server routes
+   */
+  private setupRoutes(): void {
+    // Add middleware
+    this.app.use("*", logger());
+    this.app.use("*", cors());
+
+    // Add project validation middleware
+    this.app.use(
+      `/org/${this.org}/projects/:projectName/*`,
```

Comment:
Is this middleware necessary? Looks like all the other routes just hardcode the project name, so I assume this is to return a slightly improved 404 message? If we do think its necessary, should we have a similar one for the org?

Reply:
Exactly - it's just for a better error message. I personally found the generic 404 confusing. I'll add in similar validation for the org.

Reply:
added validation for org as well. fortunately, it can all go in the same middleware function.